Update k#epic to 0.5-BETA lnd

dannypaz opened the pull request:
This PR pulls the latest changes from LND to update the k#epic changes.

Changes:
1. deleted `Gopkg.lock` and updated deps
2. deleted `rpc.pb.*` files and regenerated the proto for lnrpc
3. merge changes w/ lnd master

Issues that were found during merge:
```
both modified:   Gopkg.lock - regenerated
both modified:   chainregistry.go - updated
both modified:   channeldb/invoice_test.go - updated
both modified:   channeldb/invoices.go - updated
both modified:   config.go - updated
both modified:   htlcswitch/interfaces.go - updated
both modified:   htlcswitch/link.go - updated
both modified:   lnd.go - updated -simnet change
both modified:   lnrpc/rpc.pb.go - regenerated
both modified:   lnrpc/rpc.proto - updated
both modified:   lnrpc/rpc.swagger.json - updated
both modified:   rpcserver.go - updated
```

wpaulino committed:
config: modify private key path to be onion type agnostic

wpaulino committed:
server: update to lastest tor controller api

wpaulino committed:
tor/controller: add support for v3 onion services

In this commit, we extend our Tor controller to also support creating v3
onion services, as they are now supported by the Tor daemon. We also
refactor our existing AddOnion method to take in a config struct that
houses all of the required options to create/restore an onion service.

wpaulino committed:
docs: update v3 tor docs

wpaulino committed:
chainntnfs: ensure proper fallback to scanning tx manually

In this commit, we address a bug where it's possible that we still
attempt to manually scan for a transaction to determine whether it's
been included in the chain even after successfully checking the txindex
and not finding it there. Now, we'll short-circuit this process by
exiting early if the txindex lookup was successful but the transaction
in question was not found. Otherwise, we'll fall back to the manual
scan.

wpaulino committed:
chainntnfs: extract helper test vars and functions

In this commit, we extract some of the helper test variables and
functions into their own file and guard them under a build flag. This is
needed as some unit tests will be introduced in a future commit where
most of the same functions within the interface tests are reused. In
order to prevent these variables and functions from being exportable, we
guard them by the "debug" build tag.

wpaulino committed:
chainntnfs: guard test chain notifier under debug flag

wpaulino committed:
multi: enable txindex on miner harness

valentinewallace committed:
chainntnfs/interface_test: fix unreliable historical block ntfns test

After joining the two forked chains, it is necessary to ensure they both agree on the same best hash before proceeding to UnsafeStart the notifier.
This is because when the BitcoindClient starts, it retrieves its best known block then calls GetBlockHeaderVerbose on the hash of the retrieved block. This block could be a reorged block if JoinNodes has not completed sync. If it is the case that the best block retrieved has been reorged out of the chain, GetBlockHeaderVerbose errors because bitcoind sets the number of confirmations to -1 on reorged blocks, and the btcd rpc client panics when parsing a block whose number of confirmations is negative.

This parsing error is expected to be fixed, and as a more permanent solution chain backends should ensure that the `best block` they retrieve during startup has not been reorged out of the chain.

wpaulino committed:
macaroons: ensure path for macaroon DB exists

Co-Authored-By: Karl Ranna <karl@karlranna.com>

wpaulino committed:
docs: update path to where macaroons are stored

Co-Authored-By: Karl Ranna <karl@karlranna.com>

cfromknecht committed:
htlcswitch/link_test: increase fwdpkg gc interval

In this commit, we increase the fwdpkg gc interval
to avoid having it conflict with switch tests that
inspect forwarding packages. The current timeout is
a little too short on travis, and sporadically fails
TestChannelLinkCleanupSpuriousResponses, which was
added recently.

cfromknecht committed:
lnd_test: prevent calling Fatal in goroutine

This commit prevents an error that I've seen on travis,
wherein the test fails because a call to Fatal happens
after the test finishes. The root cause is that we call
Fatal in a goroutine that is reading from the subscribe
graph rpc call.

To fix this, we now pass an err chan back into the main
test context, where we can receive any errors and fail
the test if one comes through.

cfromknecht committed:
contractcourt/channel_arbitrator: handle onchain close race on restart

Roasbeef committed:
config: update comment for network directory

Roasbeef committed:
docker: update docker README to account for recent lncli/macaroon changes

Roasbeef committed:
cmd/lncli: properly parse cert and macaroon paths for all config variants

In this commit, we ensure that we're able to properly parse the cert and
macaroon paths for all relevant config variants. Before this commit, it
would be the case that the macaroon path ended up empty if a user wasn't
running the default (mainnet, lnddir) settings. In this commit, we
remedy this by parsing each of the two (cert+macaroon) paths
independently.

Roasbeef committed:
cmd/lncli: rename parseArgs to extractPathArgs, return error

Roasbeef committed:
Merge pull request #1754 from halseth/fndmgr-gossiper-handoff

Unify gossiper and router announcement handling

cfromknecht committed:
autopilot/agent: add OnNodeUpdates signal

Adds a new external signal alerting autopilot that
new nodes have been added to the channel graph or
an existing node has modified its channel
announcment. This allows autopilot to examine its
current state, and attempt to open channels if our
target state is not yet met.

cfromknecht committed:
pilot: signal OnNodeUpdates to autopilot

cfromknecht committed:
autopilot/agent_test: test that agent recovers from initial empty graph

Roasbeef committed:
Merge pull request #1731 from halseth/link-policy-persist

Correctly apply min_htlc to forwarding policy

Roasbeef committed:
lnwire: recognize required query sync feature bit

Roasbeef committed:
lnwallet: fix logging message for fee floor

Roasbeef committed:
config: properly show list of sub-systems for show debuglevel

Fixes #1772.

Roasbeef committed:
build: update btcd+btcwallet+neutrino to latest versions

Roasbeef committed:
chainntnfs/neutrinonotify: update to latest API changes

Roasbeef committed:
routing/chainview: update to latest API changes

Roasbeef committed:
Merge pull request #1776 from cfromknecht/autopilot-add-node-trigger

[autopilot]: add OnNodeUpdate trigger

cfromknecht committed:
chainntnfs/bitcoind: initialize with height hint cache

cfromknecht committed:
chainntnfs/btcdnotify: initialize with height hint cache

halseth committed:
chainntnfs/test_util: set 10ms trickleInterval for miner

halseth committed:
chainntnfs/bitcoind tests: check found mempool case

halseth committed:
chainntnfs/btcd: remove unnecessary check for tx==nil

halseth committed:
chainntnfs/bitcoind: remove unnecessary check for tx==nil

halseth committed:
chainntnfs: define common TxConfStatus type

halseth committed:
chainntnfs/btcd+bitcoind: use common TxConfStatus

halseth committed:
chainntnfs/btcd+bitcoind tests: use common TxConfStatus

halseth committed:
chainntnfs/btcd: fallback to scan manually only in case of err != nil

halseth committed:
chainntnfs/bitcoind: fallback to scan manually only in case of err != nil

cfromknecht committed:
chainntnfs/interface: fix TxConfStatus godocs for linter

Roasbeef committed:
Merge pull request #1778 from cfromknecht/txindex-proper-fallback

Txindex proper fallback

Roasbeef committed:
Merge pull request #1769 from cfromknecht/increase-fwdpkg-gc-test-interval

htlcswitch/test_utils: increase fwdpkg gc interval

Roasbeef committed:
Merge pull request #1764 from cfromknecht/isolate-fwdpkg-to-live-links

htlcswitch/link: only resovle+gc fwdpkgs for live channels

Roasbeef committed:
peer: raise readHandler wait group done to defer statement

In this commit, we raise the readHandler wait group done into a defer
statement at the top of the method. This fixes an existing but that
would cause the readHandler to declare it had exited, yet possibly still
be waiting on the chan message stream below to exit.

Roasbeef committed:
peer: stop chanMsg streams in defer right after creation

Roasbeef committed:
peer: move atomic var increment in msgConsumer goroutine to defer

In this commit we move the atomic var increment that signals the
consumer goourtine has exited to the top of the method in a defer
statement. This cleans up some duplicate code and also adheres to the
pattern of using defers to signal cleaning up any dependent goroutine
state on exit.

Roasbeef committed:
funding: add caller quit channel to waitUntilChannelOpen

In this commit, we add a caller quit channel to waitUntilChannelOpen.
This ensures that the caller won't block forever if it needs to exit
before the funding manager exits, or the channel barrier is actually
closed.

Roasbeef committed:
peer: allow msgStream goroutines to quit within the apply function

In this commit, we thread through the quit of the peer to the execution
of the apply function for a msgStream. This change ensures that if the
target is still processing the message, then the peer is able to exit
cleanly and not block insensately.

Roasbeef committed:
peer: allow msgConsumers to also exit on peer quit send

Roasbeef committed:
server: don't attempt to disable private channels

Roasbeef committed:
peer: don't attempt to re-enable any private channels

Roasbeef committed:
funding+peer: don't attempt to deliver messages if the peer is shutting down

Roasbeef committed:
build: update dep for golang.org/x/crypto to compile under golang 1.11

cfromknecht committed:
chainntnfs/height_hint_cache: add disable flag to hint cache

cfromknecht committed:
chainntnfs/height_hint_cache_test: add tests for disabled cache

cfromknecht committed:
chainntnfs/interface_test: run tests w/ disabled cache

cfromknecht committed:
lnwallet/interface_test: run tests with disabled hint cache

cfromknecht committed:
chainregistry: disable height hint cache

Roasbeef committed:
lnpeer: extend Peer interface with new QuitSignal method

In this commit, we extend the Peer interface with a new QuitSignal
method. This method is meant to expose a read-only quit channel which
will allow callers to cancel any actions based on the lifetime of the
underlying peer.

Roasbeef committed:
peer: add QuitSignal to ensure peer struct adheres to lnpeer.Peer interface

Roasbeef committed:
htlcswitch+funding+discovery: update mock peers to add new QuitSignal method

Roasbeef committed:
discovery: select on peer's QuitSignal to allow caller to unblock if disconnecting

In this commit, we select on the peer's QuitSignal to allow the caller
to unblock if the peer itself is disconnecting. With this change, we now
ensure that it isn't possible for a peer to block on this method and
prevent a graceful exit.

Roasbeef committed:
peer: before and after obtaining link for chan update, check quit signal

Roasbeef committed:
Merge pull request #1785 from Roasbeef/ensure-peer-quit

peer+server: ensure the peer is always able to quit even mid msgStream application

cfromknecht committed:
chainntnfs/bitcoindnotify: disable height hints in testing

cfromknecht committed:
chainntnfs/btcdnotify: disable height hint cache in testing

cfromknecht committed:
lnd_test: correct force close test re: htlc limbo funds

This commit fixes a bug in the integration test, that
reliably fails after disabling the height hint cache.
The test originally asserted that the htlc funds were
in limbo, but was reading a stale copy of the force
close information. Recently, the test was amended to
provided a valid read of the force close in
96a079873a6da2f0a93adc30945971fc07c5e610. However,
the issue was not apparent until build against the
disabled height hint cache.

The test is now correct to assert that there are no
funds in limbo, as the commitment output has been
swept, but the htlcs are still in the contract
court, so the nursery is unaware of them. We also
add another sanity check to validate that there are
no pending htlcs on the force close at that point
in time.

Roasbeef committed:
peer: ensure readHandler doesn't block on AddMsg to msgStream

In this commit, we add a quit channel to the AddMsg method of the
msgStream struct. Before this commit, if the queue was full, the
readHandler would block and be unable to exit. We remedy this by
leveraging the existing quit channel of the peer as an additional select
case within the AddMsg method.

Commit:
merge with master

Commit:
delete auto generated files

Commit:
github merge changes for lnd 0.5 merge

Roasbeef committed:
Merge pull request #1788 from cfromknecht/disable-height-hint-cache

Disable height hint cache

dannypaz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Merge upstream LND](https://trello.com/c/9q5nzTOB/462-merge-upstream-lnd)

Roasbeef committed:
lnd+cmd/lncli: bump version to 0.5-beta

wpaulino committed:
discovery/gossiper: remove optimistic channel announcement request

In this commit, we aim to resolve an issue with nodes requesting for
channel announcements when receiving a channel update for a channel
they're not aware of. This can happen if a node is not caught up with
the chain or if they receive updates for zombie channels. This would
lead to a spam issue, as if a node is not caught up with the chain,
every new update they receive is premature, causing them to manually
request the backing channel announcement. Ideally, we should be able to
detect this as a potential DoS vector and ban the node responsible, but
for now we'll simply remove this functionality.

Roasbeef committed:
Merge pull request #1798 from wpaulino/premature-chan-ann-fetch

discovery/gossiper: remove optimistic channel announcement request

cfromknecht committed:
server: stop requesting initial graph sync

In this commit, we remove signaling for initial routing
dumps, which create unnecessary log spam, bandwidth, and
CPU. Now that gossip syncing is in full force, we will
instead opt to use the more efficient querying/set
reconciliation. Other nodes may still request initial
gossip sync from us, and we will respond.

wpaulino committed:
config: ensure ZMQ options when read from the config file are not equal

wpaulino committed:
config: defer creating the base lnd dir until all flag parsing is done

In this commit, we defer creating the base lnd directory until all flag
parsing is done. We do this as it's possible that the config file
specifies a lnddir, but it isn't actually used as the directory has
already been created.

wpaulino committed:
tor/controller: add tor version number check

In this commit, we add a check for the Tor server's version number to
ensure it supports creating v3 onion services through its control port.

treygriffith reviewed (approved):
We should do some extensive testing of this change, but otherwise this looks good.

xsb committed:
lnrpc: change lightning-api repo link redirect

wpaulino committed:
autopilot: use updateBalance rather than tracking balance explicitly

In this commit, we modify the balanceUpdate autopilot signal to update
the balance according to what's returned to the WalletBalance callback
rather than explicitly tracking the balance. This gives the agent a
better sense of what the wallet's balance actually is.

wpaulino committed:
htlcswitch: return concrete error from GetLinksByInterface

wpaulino committed:
server: avoid logging error if no links are found

In this commit, we avoid logging an error when the links associated with
a peer are not found within its termination watcher. We do this to
prevent a benign log message as the links have already been removed from
the switch.

wpaulino committed:
pilot: remove disconnection logic from chanController

In this commit, we remove the disconnection logic within the
chanController when failing to open a channel with a peer. We do this as
it's already done within the autopilot agent, where it should be, and
because it's possible that we were already connected to this node and we
happened to disconnect them anyway.

wpaulino committed:
pilot: improve error when unable to reach any of a peer's addresses

Commit:
Merge branch 'master' of github.com:lightningnetwork/lnd into k#epic/updates

Roasbeef committed:
autopilot: update the Node interface to return a raw bytes, not the key

In this commit, we modify the Node interface to return a set of raw
bytes, rather than the full pubkey struct. We do this as within the
package, commonly we only require the pubkey bytes for fingerprinting
purposes. Before this commit, we were forced to _always_ decompress the
pubkey which can be expensive done thousands of times a second.

Roasbeef committed:
discovery: update autopilot.Node usage to match recent API changes

Commit:
fixed invoice for including preimage flag in db and fixed tests for siment change

cfromknecht committed:
lnd_test: add wait predicates to pending channel checks

cfromknecht committed:
routing/router: add assume chan valid

cfromknecht committed:
server: pass AssumeChannelValid to router

cfromknecht committed:
config: add experimental assumechanvalid flag

cfromknecht committed:
peer: ensure link failures are processed in peer life cycle

cfromknecht committed:
autopilot/agent: track pending connections

This commit modifies the autopilot agent to track
all pending connection requests, and forgo further
attempts if a connection is already present.
Previously, the agent would try and establish
hundreds of requests to a node, especially if the
connections were timing out and not returning.

This resulted in an OOM OMM when cranking up
maxchannels to 200, since there would be close
to 10k pending connections before the program was
terminated. The issue was compounded by periodic
batch timeouts, causing autopilot to try and
process thousands of triggers for failing
connections to the same peer.

With these fixes, autopilot will skip nodes that we
are trying to connect to during heuristic selection.
The CPU and memory utilization have been significantly
reduced as a result.

cfromknecht committed:
autopilot/interface: expose NodeID in AttachmentDirective

cfromknecht committed:
autopilot/prefattach: set NodeID for selection candidates

halseth committed:
autopilot/agent: move signal processing out of select

halseth committed:
autopilot/agent: signal nodeUpdates on own channel

We do this to avoid a huge amount of goroutines piling up on initial
graph sync, as they will all block trying to send the node update on the
stateUpdates channel. Now we instead make a new buffered channel
nodeUpdates, and just return immediately if there is already a signal in
the channel waiting to be processed.

cfromknecht committed:
autopilot/agent_test: ensure directives use unique keys

This commit ensures that the mock attachment
directives use unique keys, ensuring that they
aren't skipped due to already having pending
connection requests. The tests fail when
they're all the same since they collide
in the pendingConns map.

cfromknecht committed:
autopilot/agent_test: adds TestAgentSkipPendingConns

Adds a test asserting that the agent prevents
itself from making duplicate outstanding
connection requests to the same peer.

halseth committed:
autopilot/agent: signal chanOpenFailureUpdates on own channel

We do this to avoid a huge amount of goroutines piling up when autopilot
is trying to open many channels, as they will all block trying to send
the update on the stateUpdates channel. Now we instead send them on a
buffered channel, similar to what is done with the nodeUpdates.

halseth committed:
autopilot/agent: signal chanPendingOpenUpdates on own channel

cfromknecht committed:
server: move gossip dispatch to peer

See next commit msg for more detail.

cfromknecht committed:
peer: log disconnect to info, remove go-errors pkg

wpaulino committed:
channeldb/graph: identify edge chan id on failure

wpaulino committed:
channeldb/graph: refactor UpdateEdgePolicy to use existing db transaction

wpaulino committed:
channeldb/graph: fix off-by-one public key slice

In this commit, we fix an off-by-one error when slicing the public key
from the serialized node info byte slice. This would cause us to write
an extra byte to all edge policies. Even though the values were read
correctly, when attempting to calculate the offset of an edge's update
time going backwards, we'd always be incorrect, causing us to not
properly prune the edge update index.

wpaulino committed:
channeldb/graph: properly determine old update timestamp for an edge

In this commit, we fix a lingering issue within the edge update index
where entries were not being properly pruned due to an incorrect
calculation of the offset of an edge's last update time. Since the
offset is being determined from the end to the start, we need to
subtract all the fields after an edge policy's last update time from the
total amount of bytes of the serialized edge policy to determine the
correct offset. This was also slightly off as the edge policy included
an extra byte, which has been fixed in the previous commit.

Instead of continuing the slicing approach however, we'll switch to
deserializing the raw bytes of an edge's policy to ensure this doesn't
happen in the future when/if the serialization methods change or extra
data is included.

wpaulino committed:
channeldb/graph_test: properly check entries within edge update index

Due to entries within the edge update index having a nil value, the
tests need to be modified to account for this. Previously, we'd assume
that if we were unable to retrieve a value for a certain key that the
entry was non-existent, which is why the improper pruning bug was not
caught. Instead, we'll assert the number of entries to be the expected
value and populate a lookup map to determine whether the correct entries
exist within it.

wpaulino committed:
channeldb/graph_test: extend prune edge update index test to update edges

In this commit, we extend TestChannelEdgePruningUpdateIndexDeletion test
to include one more update for each edge. By doing this, we can
correctly determine whether old entries were properly pruned from the
index once a new update has arrived.

wpaulino committed:
channeldb: add migration to properly prune edge update index

In this commit, we introduce a migration to fix some of the recent
issues found w.r.t. the edge update index. The migration attempts to fix
two things:

1) Edge policies include an extra byte at the end due to reading an
extra byte for the node's public key from the serialized node info.

2) Properly prune all stale entries within the edge update index.

As a result of this migration, nodes will have a slightly smaller in
size channeldb. We will also no longer send stale edges to our peers in
response to their gossip queries, which should also fix the fetching
channel announcement for closed channels issue.

Roasbeef committed:
lnwire: add new compatibility parsing more for onion error chan updates

In this commit, we add a compatibility mode for older version of
clightning to ensure that we're able to properly parse all their channel
updates. An older version of c-lightning would send out encapsulated
onion error message with an additional type byte. This would throw off
our parsing as we didn't expect the type byte, and so we always 2 bytes
off. In order to ensure that we're able to parse these messages and make
adjustments to our path finding, we'll first check to see if the type
byte is there, if so, then we'll snip off two bytes from the front and
continue with parsing. if the bytes aren't found, then we can proceed as
normal and parse the request.

cfromknecht committed:
routing/conf: add experimental assume valid conf

cfromknecht committed:
Makefile: add tags env argument to build and install

Roasbeef committed:
lnwire: add new ExtraOpaqueData field to gossip messages

In this commit, we add a new field to all the existing gossip messages:
ExtraOpqueData. We do this, as before this commit, if we came across a
ChannelUpdate message with a set of optional fields, then we wouldn't be
able to properly parse the signatures related to the message. If we
never corrected this behavior, then we would violate the forwards
compatible principle we use when parsing existing messages.

As these messages can now be padded out to the max message size, we've
increased the MaxPayloadLength value for all of these messages.

Fixes #1814.

Roasbeef committed:
lnwire: extend quickcheck parser tests to add extra data for gossip messages

Roasbeef committed:
channeldb: add new ExtraOpaqueData to edge policy+update and node ann

In this commit, we add a mirror set of fields to the ones we recently
added to the set of gossip wire messages. With these set of fields in
place, we ensure that we'll be able to properly store and re-validate
gossip messages that contain a set of extra/optional fields.

Roasbeef committed:
Merge pull request #1812 from cfromknecht/wait-predicate-pending-channels

lnd_test: add wait predicates to pending channel checks

Roasbeef committed:
channeldb: extend set of graph tests to include opque data where pertinent

Roasbeef committed:
discovery: update utils to properly include opaque data in gossip msgs

Roasbeef committed:
discovery: ensure we populate ExtraOpaqueData when writing/reading from disk

Roasbeef committed:
discovery: add new set of tests to ensure we validate w/ extra data

Roasbeef committed:
chanseries+server: populate ExtraOpaqueData field when reading from disk

cfromknecht committed:
server: always backoff for unstable peers

This commit modifies the connection peer backoff
logic such that it will always backoff for "unstable"
peers. Unstable in this context is determined by
connections whose duration is shorter than 10
minutes. If a disconnect happens with a peer
whose connection lasts longer than 10 minutes,
we will scale back our stored backoff for that peer.

This resolves an issue that would result in a tight
connection loop with remote peers. This stemmed
from the connection duration being very short,
and always driving the backoff to the default
backoff of 1 second. Short connections like
this are now caught by the stable connection
threshold.

This also modifies the computation on the
backoff relaxation to subtract the connection
duration after applying randomized exponential
backoff, which offers better stability when
the connection duration and backoff are roughly
equal.

cfromknecht committed:
autopilot/agent_test: remove Fatalf calls from goroutines

Roasbeef committed:
peer: ensure we unlock the msgCond during peer msgConsumer exit

cfromknecht committed:
server: attempt reconnection to all known addresses

cfromknecht committed:
server: remove unused lightningID field

Roasbeef committed:
build: update to latest versions of neutrino+btcwallet

In this commit, we update to the latest versions of btcwallet+neutrino
that fix a number of bugs within lnd itself. Namely, we ensure that we
no longer print out garbage bytes, properly reconnect btcd after being
disconnected, ensure we don't add duplicate utxos, and finally ensure
that we always start the rescan from the wallet's initial birthday.

Fixes #1775.
Fixes #1494.
Fixes #444.

Roasbeef committed:
Merge pull request #1770 from cfromknecht/prevent-goroutine-fail

lnd_test: Prevent calling Fatal in goroutine

Roasbeef committed:
Merge pull request #1809 from wpaulino/autopilot-balance-update

autopilot/agent: use updateBalance rather than tracking balance explicitly

Roasbeef committed:
Merge pull request #1810 from wpaulino/switch-no-links-found

htlcswitch+server: avoid logging error if no links are found within peerTerminationWatcher

Roasbeef committed:
Merge pull request #1822 from Roasbeef/chan-update-compat

lnwire: add new compatibility parsing more for onion error chan updates

Roasbeef committed:
Merge pull request #1823 from cfromknecht/dont-dc-on-link-fail

peer: ensure link failures are processed in peer life cycle

Roasbeef committed:
Merge pull request #1811 from Roasbeef/autopilot-cpu-usage-fix

autopilot: modify the graph interface to return raw bytes for node pubkeys, not entire key

cfromknecht committed:
peer: dispatch gossip sync in peer start

This commit moves the gossip sync dispatch
such that it is more tightly coupled to the
life cycle of the peer. In testing, I noticed
that the gossip syncer needs to be dispatched
before the first gossip messages come across
the wire.

The prior spawn location in the server happens
after starting all of the peer's goroutines,
which could permit an ordering where the
gossip syncer has not yet been registered.
The new location registers the gossip syncer
within the read handler such that the call is
blocks before any messages are read.

cfromknecht committed:
lnwire/features: add required gq to local features

This commit adds the required feature name to our
set of local known features. This will allow other
peers connecting to us to set the required gossip
queries feature bit. This is required for the
subsequent commits, which instruct the server to
set the bit depending on user configured preferences.

halseth committed:
autopilot/agent: signal balanceUpdates on own channel

cfromknecht committed:
autopilot/multi: replace PubKey -> NodeKey on directive

Commit:
Merge branch 'master' of github.com:lightningnetwork/lnd into k#epic/updates

Roasbeef committed:
Merge pull request #1837 from Roasbeef/btcsuite-dep-catchup-5

build: update to latest versions of neutrino+btcwallet

Roasbeef committed:
autopilot: optimize heavy loaded agent by fetching raw bytes for ChannelEdge

In this commit, we implement an optimization to the autopilot agent to
ensure that we don't spin and waste CPU when we either have a large
graph, or a high max channel target for the agent. Before this commit,
each time we went to read the state of a channel from disk, we would
decompress the EC Point each time. However, for the case of the instal
ChannlEdge struct to feed to the agent, we only actually need to obtain
the pubkey, and can save the potentially expensive point decompression
for each directional channel in the graph.

Roasbeef committed:
channeldb: add limit on the max number of opaque bytes per announcement

In this commit, we add a new limit on the largest number of extra opaque
bytes that we'll allow to be written per vertex/edge. We do this in
order to limit the amount of disk space that we expose, as it's possible
that nodes may start to pad their announcements adding an additional
externalized cost as nodes may need to continue to store and relay these
large announcements.

cfromknecht committed:
discovery/gossiper: require explict gossip syncer init

This commit removes the fallback in fetchGossipSyncer
that creates a gossip syncer if one is not registered
w/in the gossiper. Now that we register gossip syncers
explicitly before reading any gossip query messages,
this should not longer be required. The fallback also
did not honor the cfg.NoChanUpdates flag, which may
have led to inconsistencies between configuration and
actual behavior.

wpaulino committed:
channeldb/graph_test: ensure policies for an edge have different
timestamps

In this commit, we ensure policies for edges we create in
TestChanUpdatesInHorizon have different update timestamps. This ensures
that there are two entries per edge in the edge update index. Because of
this, the test will fail because ChanUpdatesInHorizon will return
duplicate channel edges due to looking at all the entries within the
edge update index. This will be addressed in a future commit to allow
the set of tests to pass once again.

wpaulino committed:
channeldb/graph: dedup channel edges returned from ChanUpdatesInHorizon

In this commit, we ensure that we de-duplicate the set of channel edges
returned from ChanUpdatesInHorizon. Other subsystems within lnd use this
method to retrieve and send all the channels with updates within a time
series to network peers. However, since the method looks at the edge
update index, which can include up to two entries per edge for each
policy, it's possible that we'd send channel announcements and updates
twice, causing extra bandwidth.

cfromknecht committed:
config: rename noencryptwallet to noseedbackup

This commit renames the confusing noencryptwallet
flag to noseedbackup, since this highlights the more
crucial information of the flags behavior to the user.
The description has also been capitalized to urge
the user think twice about what they're doing.

cfromknecht committed:
docs/INSTALL: remove noencryptwallet from install docs

We will be slowly phasing this out, though abruptly
discontinuing support would be a more extensive change.
For now, we will ensure that this feature is not
recommended to users setting up a new wallet.

cfromknecht committed:
lntest/node: change noencryptwallet to noseedbackup

Roasbeef committed:
lnwallet: ensure that each backend uses a distinct seed in integration tests

In this commit, we add an additional degree of isolation to the set of
integration tests. A bug was recently fixed to ensure that the wallet
always starts rescans from _after_ it's birthday. In the past it would
miss some funds that were deposited _right_ before the birthday of the
wallet. Fixing this bug exposed a test flake wherein the btcd node would
itself rescan back and collect some of the funds that were last sent to
the bitcoind node.

In order to fix this, we now ensure that each backend will use a unique
HD seed such that the tests are still deterministic for each backend and
role.

cfromknecht committed:
lnd: use default enc pw when noseedbackup is passed

cfromknecht committed:
cmd/lncli/commands: change noencryptwallet -> noseedbackup

cfromknecht committed:
docker/lnd/start-lnd.sh: pass noseedbackup instead of noencryptwallet

cfromknecht committed:
sample-lnd.conf: remove noencryptwallet flag from sample

cfromknecht committed:
walletunlocker/service: change noencryptwallet reference to noseedbackup

cfromknecht committed:
macaroons/README: update noencryptwallet -> noseedbackup

Roasbeef committed:
Merge pull request #1840 from cfromknecht/backoff-short-conns

server: always backoff for unstable peers

Roasbeef committed:
Merge pull request #1774 from cfromknecht/onchain-close-race

contractcourt/channel_arbitrator: handle onchain close race on restart

Roasbeef committed:
Merge pull request #1836 from cfromknecht/reconnect-to-all-known-addrs

server: Reconnect to all known addrs for persistent peers

Roasbeef committed:
Merge pull request #1821 from wpaulino/prune-edge-update-index

channeldb: properly prune stale entries within edge update index

Roasbeef committed:
Merge pull request #1839 from cfromknecht/assume-channel-valid

routing: Assume channel valid

Roasbeef committed:
Merge pull request #1842 from wpaulino/dedup-chan-updates-horizon

channeldb: dedup channel edges returned from ChanUpdatesInHorizon

cfromknecht committed:
discovery/syncer: delay replies after initial sync to prevent DOS

cfromknecht committed:
discovery/gossiper: remove go-errors pkg

roeierez committed:
amount paid should be in satoshies as other fields used for amount

cfromknecht committed:
discovery/syncer_test: add DOS delayed replies test

halseth committed:
Merge pull request #1846 from roeierez/invoice_amt_paid

rpcserver: amount paid should be in satoshis as other fields used for amount

cfromknecht committed:
dep: constrain siphash, x/sys, and errgo.v1, upd neutrino

Roasbeef committed:
Merge pull request #1841 from cfromknecht/sync-gq-start

peer: Synchronous registration of gossip queries

Roasbeef committed:
channeldb: also ignore the EOF error when trying to read ExtraOpaqueBytes

In this commit, we account for the additional case wherein the
announcement hasn't yet been written with the extra zero byte to
indicate that there aren't any remaining bytes to be read. Before this
commit, we accounted for the case where the announcement was written
with the extra byte, but now we ensure that legacy nodes that upgrade
will be able to boot properly.

Roasbeef committed:
Merge pull request #1802 from wpaulino/equal-zmq-conf-error

config: ensure ZMQ options are not equal and defer creating lnddir until all flag parsing is done

Roasbeef committed:
Merge pull request #1825 from Roasbeef/extra-gossip-message-data

channeldb+discovery: ensure we store, validate and propagate announcements with opaque data

Roasbeef committed:
Merge pull request #1849 from cfromknecht/resync-golock

dep: constrain siphash, x/sys, and errgo.v1 + upd neutrino

Roasbeef committed:
Merge pull request #1817 from halseth/autopilot-node-update

Reduce goroutine overhead on autopilot updates

wpaulino committed:
peer: prune persistent peer connection on zero on-disk channels

In this commit, we fix a small bug with regards to the persistent peer
connection pruning logic. Before this commit, it'd be the case that we'd
prune a persistent connection to a peer if all links happen to be
inactive. This isn't ideal, as the channels are still open, so we should
always be atttempting to connect to them. We fix this by looking at the
set of channels on-disk instead and prune the persistent connection if
there aren't any.

Roasbeef committed:
Merge pull request #1844 from wpaulino/prune-peer-conn-disk-chans

peer: prune persistent peer connection on zero on-disk channels

Roasbeef committed:
lnd: ensure that we set the wallet birthday for --noencryptwallet

In this commit, we ensure that we always set the wallet birthday. If the
user has provided a seed, or is creating a new one, then it will be
overwritten below. However, before this commit, if a user started with
the --noencryptwallet flag, then we would _always_ start to rescan from
genesis with the recent bug fix to ensure that we always start after the
birthday.

Roasbeef committed:
Merge pull request #1801 from cfromknecht/remove-initial-graph-sync-request

server: Stop requesting initial graph sync

cfromknecht committed:
lnwire/lnwire_test: assert nil 0-length sid reply msg

Roasbeef committed:
Merge pull request #1818 from cfromknecht/aplt-track-pending-conns

autopilot: prevent duplicate conns to same peer

Roasbeef committed:
Merge pull request #1740 from valentinewallace/fix-historical-ntfns-reorg-test

chainntnfs/interface_test: fix unreliable historical block ntfns test

halseth committed:
lnrpc: add field amt_paid_msat to Invoice

halseth committed:
rpcserver: set AmtPaidMsat when creating RPC invoice

Commit:
merge with master

wpaulino committed:
server: ensure persistent connections to tor addrs are made if active

In this commit, we restrict the persistent connection logic on startup
to only attempt to establish connections to Tor addresses if Tor
outbound support is enabled. Otherwise, we'll continually attempt to
reach the address even though we never will.

Roasbeef committed:
chanseries: fix panic bug by ensuring we access a non-nil edge

In this commit, we fix an existing but that would cause the daemon to at
times crash. Before this commit, we access the wrong edge, which would
possibly actually be nil, leading to a panic. In this commit we fix this
by ensuring we access the proper edge which is known to be non-nil at
this point in the control flow.

Roasbeef committed:
lntest: log the expected and last balance in WaitForBalance error case

Roasbeef committed:
test: fix flake in link node garbage collection test

In this commit, we fix a flake in the link node garbage collection test
by ensuring the channels have been fully closed on both sides before we
attempt to restart and ensure that they don't actually establish
connections. Without this check, it's possible that either side hasn't
yet processed all the blocks, so they'll still reconnect to each other on
start up.

Roasbeef committed:
build: update to latest version of neutrino

In this commit we update to the latest version of neutrino which fixes a
bug in the filter header syncing logic.

Roasbeef committed:
Merge pull request #1862 from Roasbeef/chan-series-test-fixes

multi: fix panic bug in chan series, itest test flake, update neutrino to latest version

Roasbeef committed:
Merge pull request #1853 from cfromknecht/short-chan-id-compare

lnwire/lnwire_test: assert nil 0-length sid reply msg

Roasbeef committed:
build: update btcwallet to latest version

In this commit, we update btcwallet to the latest version which fixes a
panic bug when attempting to notify a relevant transaction.

Roasbeef committed:
Merge pull request #1864 from Roasbeef/btcwallet-update-panic-fix

build: update btcwallet to latest version

Roasbeef committed:
cmd/lncli+fundingmanager: use a default num min confs for 1 for openchannel

Roasbeef committed:
Merge pull request #1855 from halseth/amt-paid-msat

Add amt_paid_msat to invoice

Roasbeef committed:
Merge pull request #1861 from wpaulino/connect-tor-if-active

server: ensure persistent connections to tor addrs are made if active

michael1011 committed:
added missing spaces to Makefile

grunch committed:
cmd/lncli: fix missing whitespaces on lncli queryroutes desc

Adds whitespaces missing on command description,
without those whitespaces at the end of the line
it showed the words together

Roasbeef committed:
Merge pull request #1824 from cfromknecht/gossip-delay-reply

discovery/syncer: delay replies after initial sync to prevent DOS

Roasbeef committed:
channeldb: remove unused fwdPackageLogBucket key

In this commit, we move the declaration of the key for an unused bucket.
In the past, this bucket was used to store the revocation forwarding
package log. However, this has been moved under the key
`fwdPackagesKey`.

Roasbeef committed:
channeldb: fix bucket creation hierarchy in createChannelDB

In this commit, we fix a bug in the bucket creation code in
createChannelDB. This bug can cause migrations on older nodes to fail,
as we expect the bucket to already have been created. With this commit,
we ensure that all the buckets under the main node and edges bucket are
properly created. Otherwise, a set of the newer migrations will fail to
apply for nodes updating from 0.4.

Commit:
merge with master

davidknezic committed:
release: Add ARMv6 platform to release.sh

wpaulino committed:
lnrpc: add fields to paginate invoices backwards

wpaulino committed:
channeldb/invoices: extend invoice queries to allow backwards pagination

wpaulino committed:
channeldb/invoice_test: refactor TestQueryInvoices and add reverse test cases

wpaulino committed:
rpcserver: include new proto fields in ListInvoices rpc

wpaulino committed:
cmd/lncli: add description and reversed flag to listinvoices

cfromknecht committed:
dep: add golang.org/x/time for rate limiter

cfromknecht committed:
discovery/syncer: use rate limiter for gossip queries

This commit replaces the simplistic rate limiting
technique added in 557cb6e2, to use the
golang.org/x/time's rate limiter. This has the
benefit of performing traffic shaping to meet a
target maximum rate, and yet tolerate bursts. Bursts
are expected during initial sync, though should become
more rare afterwards. Performing traffic shaping with
this mechanism should improve the ability of the gossip
syncer to detect sustained bursts from the remote peer,
and penalize them appropriately.

This commit also modifies the default parameters to
accept bursts of 10 queries, with a target rate of 1
reply every 5 seconds.

halseth committed:
lnd_test: restart nodes twice to trigger pruning logic

halseth committed:
lnd_test: make sure we stay disconnected using WaitInvariant in testGarbageCollectLinkNodes

tsarpaul committed:
cmd/lncli: Fixed conflicting payinvoice amount output

Roasbeef committed:
Merge pull request #1872 from Roasbeef/create-proper-buckets

channeldb: fix bucket creation hierarchy in createChannelDB

cfromknecht committed:
contractcourt/chain_arbitrator: parallel chain watcher start

This commit restructures the initialization procedure
for chain watchers such that they can proceed in parallel.
This is primarily to help nodes running with the neutrino
backend, which otherwise forces a serial rescan for each
active channel to check for spentness.

Doing so allows the rescans to take advantage of batch
scheduling in registering for the spend notifications,
ensuring that only one or two passes are made, as opposed
to one for each channel.

Lastly, this commit ensures that the chain arb is properly
shutdown if any of it's chain watchers or channel arbs
fails to start, so as to cancel their goroutines before
exiting.

Roasbeef committed:
channeldb: relax bucket assumptions for latest db migration

In this commit, we no longer assume that the bucket hierarchy has been
created properly when applying the latest DB migration. On older nodes
that never obtained a channel graph, or updated _before_ the query sync
stuff was added, then they're missing buckets that the migration expects
them to have.

We fix this by simply creating the buckets as we go, if needed.

Roasbeef committed:
Merge pull request #1877 from halseth/garbage-collect-link-nodes

[integration tests] Fix flake in persistent connection pruning test

Roasbeef committed:
Merge pull request #1886 from Roasbeef/lax-migration-assumptions

channeldb: relax bucket assumptions for latest db migration

Roasbeef committed:
Merge pull request #1887 from cfromknecht/parallel-chain-watcher-start

contractcourt/chain_arbitrator: parallel chain watcher start

Roasbeef committed:
Merge pull request #1880 from cfromknecht/rate-limit-gossip-queries

discovery: Traffic shaping of gossip query replies

Roasbeef committed:
Merge pull request #1878 from davidknezic/release-armv6

release: Add ARMv6 platform to release.sh

Roasbeef committed:
Merge pull request #1516 from wpaulino/auto-tor-v3

tor+server: add automatic support for v3 onion services

wpaulino committed:
multi: update to new BitcoindClient API

Roasbeef committed:
build: update dep to latest version of btcwallet

Roasbeef committed:
lnwallet: add test to ensure that all backends show unconf txns in ListTransactionDetails

In this commit, we add a new test to the set of lnwallet integration
tests. In this new test, we aim to ensure that all backends are able to
display unconfirmed transactions in ListChainTransactions. As of this
commit, this test fails as no backends will return unconfirmed
transactions properly.

Roasbeef committed:
lnwallet: fix ListTransactionDetails to ensure unconfirmed transactions are installed

In this commit, we fix a bug in the arguments to GetTransactions for the
btcwallet implementation of the WalletController interface. Before this
commit, we wouldn't properly return unconfirmed transactions. The issue
was that we didn't specify the special mempool height of "-1", as the
ending height. The mempool height is actually internally converted to
the highest possible height that can fit into a int32.

In this commit, we set the start to zero, and end to -1 (actually
2^32-1) to properly scan for unconfirmed transactions.

Fixes #1422.

Roasbeef committed:
lnwallet: update for new API change in NewBitcoindClient

Roasbeef committed:
lnwallet: add test to execrise unconfirmed transcations notified by SubscribeTransactions

In this commit, we add a new test to ensure that all backends will
properly send out notifications when an unconfirmed transcation that we
send is inserted into the tx store. Before we updated the btcwallet
build commit in dep, this would fail for neutrino but now passes.

Roasbeef committed:
Merge pull request #1843 from cfromknecht/change-noencryptwallet

Start deprecating noencryptwallet

Roasbeef committed:
config: return an error if unable to parse values in the config file

Roasbeef committed:
channeldb: don't use KeyN in latest migration

In this commit, we fix a bug in the latest migration that could cause
the migration to end in a panic. Additionally, we modify the migration
to exit early if the bucket wasn't found, as in this case, no migration
is required.

Fixes #1874.

Roasbeef committed:
config: fix linter error

Roasbeef committed:
Merge pull request #1896 from Roasbeef/btcwallet-unconf-txn

lnwallet+rpc: allow unconfirmed transactions to be returned over RPC, fix SubscribeTransaction unconf for neutrino

Roasbeef committed:
Merge pull request #1863 from wpaulino/query-invoices-default

channeldb+cmd/lncli: enable backwards pagination of listinvoices

Roasbeef committed:
Merge pull request #1898 from Roasbeef/fix-migration

channeldb: don't use KeyN in latest migration

cfromknecht committed:
make: clean lnd-debug, lncli-debug and .vendor-new

wpaulino committed:
cmd/lncli: update listinvoices docs with new pagination support

wpaulino committed:
lnrpc: update ListInvoices RPC with new pagination support

Roasbeef committed:
discovery: fix logging for gossip syncer rate limiting

Roasbeef committed:
lnwire: check length of payload in decodeShortChanIDs

Roasbeef committed:
README: add security section in readme

Roasbeef committed:
build: update to point to latest version of neutrino

In this commit, we update the build to point to the latest version of
neutrino which includes a bug fix for a regression that would cause the
daemon to spin when at chain tip attempting to always fetch the next set
of headers though it was already fully up to date.

Roasbeef committed:
Merge pull request #1869 from michael1011/makefile-space

Added missing spaces to Makefile

Roasbeef committed:
Merge pull request #1885 from tsarpaul/master

cmd/lncli: Fixed conflicting payinvoice amount output

Roasbeef committed:
Merge pull request #1901 from cfromknecht/clean-vendor-new

make: clean lnd-debug, lncli-debug and .vendor-new

Roasbeef committed:
channeldb: fix bug in migration from 0.4 to 0.5

In this commit, we fix a bug in the latest database migration when
migrating from 0.4 to 0.5. There's an issue in bolt db where if one
deletes a bucket that has a key with a nil value, it thinks that's a sub
bucket and attempts a bucket deletion. This will fail as it's not
actually a sub-bucket.  We get around this by using a cursor to manually
delete items in the
bucket.

Fixes #1907.

Commit:
updated lock and rpc files

dannypaz commented:
@treygriffith closing this branch in favor of https://github.com/sparkswap/lnd/pull/14 because I had issues w/ integration tests and it was easier to start over w/ the official release
